fabtests/pytest/common.py

```diff
@@ -33,17 +34,31 @@ def has_ssh_connection_err_msg(output):
 
     return False
 
-
+@functools.lru_cache(10)
 @retry(retry_on_exception=is_ssh_connection_error, stop_max_attempt_number=3, wait_fixed=5000)
-def has_cuda(ip):
-    outfile = NamedTemporaryFile(prefix="nvidia_smi.").name
-    proc = run("ssh {} nvidia-smi -L > {} 2>&1".format(ip, outfile), shell=True)
-    output = open(outfile).read()
-    os.unlink(outfile)
-    if has_ssh_connection_err_msg(output):
+def num_cuda_devices(ip):
+    proc = run("ssh {} nvidia-smi -L".format(ip), shell=True,
+               stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+               timeout=60, encoding="utf-8")
+
+    if has_ssh_connection_err_msg(proc.stderr):
         raise SshConnectionError()
 
-    return proc.returncode == 0
+    # the command "nvidia-smi -L" print 1 line for each GPU
+    # An example line is like:
+    #  GPU 0: NVIDIA A100-SXM4-40GB (UUID: GPU-ddba3c80-ed95-c778-0c47-8cdf2ce99787)
+    # Therefore here we count number of lines that starts with GPU
+    result = 0
+    lines = proc.stdout.split("\n")
+    for line in lines:
+        if line.find("GPU") == 0:
+            result += 1
```

Comment:
I suppose if there is no nvidia-smi, the logic here will try to search "nvidia-smi: command not found" and just return 0 right?  looks good to me.

Reply:
Yes.